samples/subsys/usb/hid-keyboard/src/main.c

```diff
@@ -0,0 +1,247 @@
+static void kb_set_protocol(const struct device *dev, const uint8_t proto)
+{
+	LOG_INF("Protocol changed to %s",
+		proto == 0U ? "Boot Protocol" : "Report Protocol");
+}
+
+static void kb_output_report(const struct device *dev, const uint16_t len,
+			     const uint8_t *const buf)
+{
+	LOG_HEXDUMP_DBG(buf, len, "o.r.");
+	kb_set_report(dev, 2U, 0U, len, buf);
+}
+
+struct hid_device_ops kb_ops = {
+	.iface_ready = kb_iface_ready,
+	.get_report = kb_get_report,
+	.set_report = kb_set_report,
+	.set_idle = kb_set_idle,
+	.get_idle = kb_get_idle,
+	.set_protocol = kb_set_protocol,
+	.output_report = kb_output_report,
+};
+
+int main(void)
+{
+	struct usbd_contex *sample_usbd;
+	const struct device *hid_dev;
+	int ret;
+
+	for (unsigned int i = 0; i < ARRAY_SIZE(kb_leds); i++) {
+		if (kb_leds[i].port == NULL) {
+			continue;
+		}
+
+		if (!gpio_is_ready_dt(&kb_leds[i])) {
+			LOG_ERR("LED device %s is not ready", kb_leds[i].port->name);
+			return -EIO;
+		}
+
+		ret = gpio_pin_configure_dt(&kb_leds[i], GPIO_OUTPUT_INACTIVE);
+		if (ret != 0) {
+			LOG_ERR("Failed to configure the LED pin, %d", ret);
+			return -EIO;
+		}
+	}
+
+	hid_dev = DEVICE_DT_GET_ONE(zephyr_hid_device);
+	if (!device_is_ready(hid_dev)) {
+		LOG_ERR("HID Device is not ready");
+		return -EIO;
+	}
+
+	ret = hid_device_register(hid_dev,
+				  hid_report_desc, sizeof(hid_report_desc),
+				  &kb_ops);
+	if (ret != 0) {
+		LOG_ERR("Failed to register HID Device, %d", ret);
+		return ret;
+	}
+
+	sample_usbd = sample_usbd_init_device(NULL);
+	if (sample_usbd == NULL) {
+		LOG_ERR("Failed to initialize USB device");
+		return -ENODEV;
+	}
+
+	ret = usbd_enable(sample_usbd);
+	if (ret) {
+		LOG_ERR("Failed to enable device support");
+		return ret;
+	}
+
+	while (true) {
+		struct kb_event kb_evt;
+
+		k_msgq_get(&kb_msgq, &kb_evt, K_FOREVER);
+
+		switch (kb_evt.code) {
+		case INPUT_KEY_0:
+			if (kb_evt.value) {
+				report[KB_KEY_CODE1] = HID_KEY_NUMLOCK;
+			} else {
+				report[KB_KEY_CODE1] = 0;
+			}
+
+			break;
+		case INPUT_KEY_1:
+			if (kb_evt.value) {
+				report[KB_KEY_CODE2] = HID_KEY_CAPSLOCK;
+			} else {
+				report[KB_KEY_CODE2] = 0;
+			}
+
+			break;
+		case INPUT_KEY_2:
+			if (kb_evt.value) {
+				report[KB_KEY_CODE3] = HID_KEY_SCROLLLOCK;
+			} else {
+				report[KB_KEY_CODE3] = 0;
+			}
+
+			break;
+		case INPUT_KEY_3:
+			if (kb_evt.value) {
+				report[KB_MOD_KEY] = HID_KBD_MODIFIER_RIGHT_ALT;
+				report[KB_KEY_CODE4] = HID_KEY_1;
+				report[KB_KEY_CODE5] = HID_KEY_2;
+				report[KB_KEY_CODE6] = HID_KEY_3;
+			} else {
+				report[KB_MOD_KEY] = HID_KBD_MODIFIER_NONE;
+				report[KB_KEY_CODE4] = 0;
+				report[KB_KEY_CODE5] = 0;
+				report[KB_KEY_CODE6] = 0;
+			}
+
+			break;
+		default:
+			LOG_INF("Unrecognized input code %u value %d",
+				kb_evt.code, kb_evt.value);
+			continue;
+		}
+
+		if (!kb_ready) {
+			LOG_INF("USB HID device is not ready");
+			continue;
+		}
+
+		ret = hid_device_submit_report(hid_dev, sizeof(report), report, true);
```

Comment:
By default, main thread is preemptive. USB HID device readiness may change after the `!kb_ready` check (it could also bypass internal checks done by `hid_device_submit_report` in similar way). Is it safe or could it lead to problems or edge cases in USB stack? Maybe we need cooperative priority here

Reply:
Why do you think the cooperative thread can not be interrupted and is somehow different then?

Reply:
A preemptive thread can be supplanted at any time by the scheduler while a cooperative thread needs to perform an action that would make it unready. As long as `input_cb` is called from a thread context (e.g. not from an interrupt), it would not preempt execution of a cooperative thread.

Reply:
cooperative thread can be interrupted by an interrupt, AFAIK there is no guaranty it can immediately continue after, am I wrong?

Reply:
Cooperative thread will continue execution after it is interrupted by an interrupt. There is guarantee that there won't be other thread executing, see https://docs.zephyrproject.org/3.6.0/kernel/services/scheduling/index.html#cooperative-time-slicing for more information.

Reply:
The `CONFIG_INPUT_MODE` is `CONFIG_INPUT_MODE_THREAD` by default, so that events seem broadcasted through a common input thread by default (then cooperative main priority could actually prevent supplant).

Ref: https://docs.zephyrproject.org/latest/services/input/index.html#application-api

Reply:
I thought synchronous is on by default, added `CONFIG_INPUT_MODE_SYNCHRONOUS=y`.

Reply:
> Cooperative thread will continue execution after it is interrupted by an interrupt. There is guarantee that there won't be other thread executing,

It seems to me that this article does not pay attention to meta-IRQ threads.

Reply:
I wonder if using a thread for synchronizing inputs wouldn't actually simplify synchronization here (we could rely on cooperative priorities to simplify sample's implementation and prevent races). You are right that meta-IRQ threads are tricky use-case here (still, I don't recall seeing them used instead of regular threads in most of the subsystems or samples - I guess using them would require some "special handling" to ensure safety)

Reply:
Still, since it's an improvement to the sample code, I would suggest to add it (but I would not NACK PR because of it)